File: PA2/Lex.c
//Lynne Diep                                                                                                                                                                         
//lytdiep@ucsc.edu                                                                                                                                                                   
//CMPS 101 - PA2                                                                                                                                                                     
//Lex.c                                                                                                                                                                              

#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include"List.h"

#define MAX_LEN 160

int main(int argc, char* argv[]){

  int n, count=0;
  FILE *in, *out;
  char line[MAX_LEN];
  char tokenlist[MAX_LEN];
  char* token;

  if( argc != 3 ){ //check that there are not 3 command line arguments                                                                                                               
    printf("Usage: %s <input file> <output file>\n", argv[0]);
    exit(1);
  }

  in = fopen(argv[1], "r");//open files for reading and writing                                                                                                                      
  out = fopen(argv[2], "w");
  if( in==NULL ){
    printf("Unable to open file %s for reading\n", argv[1]);
    exit(1);
  }
  if( out==NULL ){
    printf("Unable to open file %s for writing\n", argv[2]);
    exit(1);
  }

  while( fgets(line, MAX_LEN, in) != NULL)  {//count number of lines                                                                                                                 
    count++;
  }

  rewind(in);//back to front of list                                                                                                                                                 

  char word[count - 1][MAX_LEN]; //copy file into string array                                                                                                                       
  n = -1;

  while(fgets(line, MAX_LEN, in) != NULL) {
    strcpy(word[++n], line);
  }

  List list = newList();

  append(list,0);

  for(int j = 1; j<count; j++){ //insertion sort                                                                                                                                     
    char*temp = word[j];
    int i = j-1;
    moveBack(list);

    while(i>=0 && strcmp(temp,word[get(list)])<=0){
      i--;
      movePrev(list);
    }
    if(index(list) >=0){
      insertAfter(list,j);
    }
    else{
      prepend(list,j);
    }
  }

  moveFront(list);

  while(index(list)>=0){ //print out sorted list                                                                                                                                     
    fprintf(out, "%s", word[get(list)],"\n");
    moveNext(list);
  }

  fclose(in); //close in file                                                                                                                                                        
  fclose(out);//close out file                                                                                                                                                       
  freeList(&list);
  return(0);
}

File: PA1/Lex.java
// Lynne Diep                                                                                                                                                                        
// lytdiep@ucsc.edu                                                                                                                                                                  
// CMPS 101                                                                                                                                                                          
// PA1                                                                                                                                                                               
// Lex.java                                                                                                                                                                          

import java.io.*;
import java.util.Scanner;

public class Lex{
    public static void main(String[] args) throws IOException{
        Scanner in = null;
        PrintWriter out = null;
        String line = null;
        String[] token = null;
        int lineNumber = 0;

        if(args.length < 2){                                                    //check that there are 2 command line arguments                                                      
            System.out.println("Usage: There are not 2 command line arguments");
            System.exit(1);
        }

        if(args.length > 2){
            System.out.println("Usage: FileIO infile outfile");
            System.exit(1);
        }

        in = new Scanner(new File(args[0]));                                   //get file                                                                                            

        while(in.hasNextLine() ) {                                             //count # of lines                                                                                    
            lineNumber++;
            in.nextLine();
        }

        in.close();                                                            //close file                                                                                          

        List list = new List();                                                //create new list                                                                                     
        token = new String[lineNumber];                                        //create array of length lineNumber                                                                   
        in = new Scanner(new File(args[0]));                                   //get file                                                                                            
        out = new PrintWriter(new FileWriter(args[1]));

        for(int i = 0; i < lineNumber; i++) {                                  //read in the lines of file, place in array                                                           
            line = in.nextLine()+" ";                                          //add extra space so split works right                                                                
            token[i] = line;
        }

        list.append(0);

        for(int j = 1; j < token.length; j++){                                 //insertion sort                                                                                      
            String key = token[j];
            int i = j-1;
            list.moveBack();

         while(key.compareTo(token[list.back()]) <=  0 && i >= 0){           //compare each line of list                                                                          
                i--;
                list.movePrev();
            }
            if(list.index() >= 0){
                list.insertAfter(j);
            }
            if(list.index() < 0){                                              //fell off list                                                                                       
                list.prepend(j);
            }
        }

        list.moveFront();

        while(list.index() >= 0){                                             //print out sorted list                                                                                
            out.println(token[list.get()]);
            list.moveNext();
        }

        in.close();                                                           //close in file                                                                                        
        out.close();                                                          //close out file                                                                                       

    }//closes IOException                                                                                                                                                            

}// closes public class Lex                          
